source/generate_cids/generate_context_groups.py

```diff
@@ -0,0 +1,135 @@
+    'http://www.nlm.nih.gov/research/umls': 'UMLS',
+    'http://pubchem.ncbi.nlm.nih.gov': 'PUBCHEM_CID',
+    'http://braininfo.rprc.washington.edu/aboutBrainInfo.aspx#NeuroNames': 'NEU',
+    'http://www.itis.gov': 'ITIS_TSN',
+}
+
+if __name__ == '__main__':
+
+    logging.basicConfig(level=logging.INFO)
+
+    ftp_host = 'medical.nema.org'
+    logger.info('log into FTP server "{}"'.format(ftp_host))
+    ftp = ftplib.FTP(ftp_host, timeout=60)
+    ftp.login('anonymous')
+
+    ftp_path = 'medical/dicom/resources/valuesets/fhir/json'
+    logger.info('list files in directory "{}"'.format(ftp_path))
+    ftp_files = ftp.nlst(ftp_path)
+    ftp_url = 'ftp://{host}/{path}'.format(host=ftp_host, path=ftp_path)
+
+    cid_pattern = re.compile('^dicom-cid-([0-9]+)-[a-zA-Z]+')
+    for ftp_filepath in ftp_files:
+        src_code_lines = []
+        ftp_filename = os.path.basename(ftp_filepath)
+        logger.info('retrieve file "{}"'.format(ftp_filename))
+        with BytesIO() as fp:
+            ftp.retrbinary('RETR {}'.format(ftp_filepath), fp.write)
+            content = fp.getvalue()
+            value_set = json.loads(content)
+
+        cid_match = cid_pattern.search(value_set['id'])
+        cid = cid_match.group(1)
+        context_group_name = value_set['name']
+        logger.info('create source code for context group CID {}'.format(cid))
+        docstring_lines = [
+            '"""CID {} {}'.format(cid, context_group_name),
+            'auto-generated by {}.'.format(os.path.basename(__file__)),
+            '"""',
+        ]
+        src_code_lines.extend(docstring_lines)
+        import_lines = [
+            'from pydicom.sr.value_types import CodedConcept',
+        ]
+        src_code_lines.extend(import_lines)
+        # Two empty lines between import statements and rest of the code.
+        # The second empty line will be added below.
+        src_code_lines.append('')
+
+        keys = set()
+        for group in value_set['compose']['include']:
+            system = group['system']
+            try:
+                scheme_designator = FHIR_SYSTEM_TO_DICOM_SCHEME_DESIGNATOR[system]
+            except KeyError:
+                raise NotImplementedError(
+                    'The DICOM scheme designator for the following FHIR system '
+                    'has not been specified: {}'.format(system)
+                )
+            for concept in group['concept']:
+                name = concept['display'].strip().upper()
+                name = name.replace(' ', '_')
+                name = re.sub(r'[-]', '', name)
+                name = re.sub(r'[^a-zA-Z0-9]', '_', name)
+                name = re.sub(r'(_+?)\1+', r'\1', name)
+                # Python variables must not begin with a number.
+                if re.match(r'[0-9]', name):
+                    name = re.sub(r'^([^_]+)(_)(.*)', r'\3_\1', name)
+                    name = re.sub(r'^([0-9]+)(.*)', r'\2_\1', name)
+                if name in keys:
```

Comment:
Need to add more rules here to account for such things as `* / ( ) '`. This table is a good test: http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_CID_7180.html. I came across this since the converter script reported repeated entries, but I think they were due to removal of those characters. There should be no duplicates in a single CID, so I suggest the warning below should be replaced with error.

Reply:
Looking at `_cid_dict.py`, it contains duplicate entries, which should absolutely not be happening - the warning should be replaced with the error, so that the person generating `_cid_dict.py` is alerted to this issue and is forced to investigate the reason.